### ISSUE TITLE
Grappler `ArgMax`/`ArgMin` monotonic stripping is wrong for strictly-monotonic ops with float32 saturation/overflow

### ISSUE THREAD

**Issue by wuyii8941** [tensorflow/tensorflow#118675]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TensorFlow 2.22.0-dev20260429

### Custom code

Yes

## Summary

Grappler's `ArithmeticOptimizer` transforms `ArgMax(f(x))` into `ArgMax(x)` for ops classified as strictly monotonic (tanh, exp, sigmoid). While these functions are mathematically strictly monotonic, float32 representation limits cause them to produce **tied** outputs (saturation to 1.0, overflow to inf, underflow to 0), changing the argmax result.

This is distinct from [#116436](https://github.com/tensorflow/tensorflow/issues/116436), which covers non-strictly-monotonic ops (relu, floor, ceil). That issue explicitly states: *"The remaining ops (Exp, Log, Sigmoid, Tanh, Softplus, etc.) are strictly monotonic and the optimization is valid for them."* This bug shows the optimization is also invalid for those ops when float32 limits create ties.

## Scope

Affects all strictly-monotonic ops that saturate or overflow in float32:

| Op | Mechanism | Example | Eager | Grappler |
|---|---|---|---|---|
| `tanh` | saturation to ±1.0 | `argmax(tanh([10, 20, 5]))` | **0** | 1 |
| `exp` | overflow to inf | `argmax(exp([100, 200, 50]))` | **0** | 1 |
| `sigmoid` | saturation to 1.0 | `argmax(sigmoid([20, 40, 5]))` | **0** | 1 |
| `exp` | underflow to 0 | `argmin(exp([-100, -200, -50]))` | **0** | 1 |
| `tanh` | saturation to -1.0 | `argmin(tanh([-5, -10, -20]))` | **1** | 2 |

The pattern is consistent: when two or more inputs produce the same float32 output (ties), Grappler's optimization returns the wrong index because it computes argmax on the original values (no ties) instead of the transformed values (with ties).

Both argmax and argmin are affected.

## Practical Impact

Temperature-scaled classification: `argmax(sigmoid(logits/T))` where logits have large magnitude or T is small. The sigmoid saturates, creating ties, and Grappler silently returns the wrong class.

Feature selection: `argmax(tanh(features))` where features have large magnitude. Common when features are not normalized.

## Root Cause

In `tensorflow/core/grappler/op_types.cc`, `IsElementWiseMonotonic()` returns `true` for tanh, exp, sigmoid, etc. The `OptimizeMaxOrMinOfMonotonic` pass in `ArithmeticOptimizer` uses this to strip the op from argmax/argmin. The transformation `ArgMax(f(x)) → ArgMax(x)` requires that `f(a) > f(b) ⟺ a > b`, which holds for strictly monotonic functions over the reals, but **not** in float32 when saturation/overflow maps distinct values to the same float32 output.

The fix in [#116709](https://github.com/tensorflow/tensorflow/pull/116709) for #116436 only addressed non-strictly-monotonic ops (relu, floor, etc.) and explicitly kept tanh/exp/sigmoid/etc. as "valid" for stripping.

## Suggested Fix

The optimization `ArgMax(f(x)) → ArgMax(x)` is only safe when `f` is injective over the representable float range. For ops that saturate (tanh, sigmoid) or overflow (exp), the optimization should be removed or restricted to inputs known to be within the injective range.

Concretely, remove `Tanh`, `Sigmoid`, `Exp`, and any other saturating/overflowing op from the list of ops eligible for argmax/argmin stripping in `IsElementWiseMonotonic()`.

## Environment

- TensorFlow 2.22.0-dev20260429
- Grappler `arithmetic_optimization` is enabled by default
- Eager mode: correct
- `arithmetic_optimization=False`: correct
- Affects both CPU and GPU (Grappler is device-independent)

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

x = tf.constant([[10.0, 20.0, 5.0]])
# tanh(x) = [1.0, 1.0, 0.9999] → argmax = 0 (first 1.0, tie-breaking)
# x       = [10, 20, 5]         → argmax = 1 (20 is largest)

print("eager:", tf.math.argmax(tf.math.tanh(x), axis=-1).numpy())  # [0]

tf.config.optimizer.set_experimental_options({"arithmetic_optimization": True})
print("Grappler:", tf.function(
    lambda x: tf.math.argmax(tf.math.tanh(x), axis=-1))(x).numpy())  # [1] WRONG


Observed on TensorFlow 2.22.0-dev20260429.
```

### Relevant log output

```shell

```

**Comment by dhantule** [comment]
Hi @wuyii8941, 
Thanks for reporting this and sharing the details.
We’ve received the issue and are taking a look. We’ll follow up here once we have more information or if we need any additional inputs.
Appreciate your patience.
